### ISSUE TITLE
dart2js: unnecessary return statement in setter

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#35918]
This setter 'returns' the value of the assignment:
```dart
   set data(Data value) => _data = value;
```
The generated code literally does the same:
```js
   set$data: function(value) {
      this._data = value;
      return value;
    }
```
AFAIK setters can never be called in a manner that gives access to the returned value, so the return statement is useless.

@munificent I have always wanted tear-offs for setters which would expose the returned value, but they could be defined to be `void` and always return `null`. What is the language direction here?

@kmillikin Should Kernel convert the `=>` to a statement for us, or is this observable on the VM, e.g. via mirrors? (It is simple enough for dart2js to deal with it.)

@johnniwinther It is an interesting global property of a function that the return value is never used because we can disconnect the return from the value expression and let dead code elimination nibble away the effect- and exception-free part of the expression and its dependencies.

**Comment by munificent** [comment]
> AFAIK setters can never be called in a manner that gives access to the returned value, so the return statement is useless.

You're correct. I think in the context of a setter, a `=>` member body should be treated like it doesn't have a return when you generate JS.

> What is the language direction here?

I don't think there's been any discussion about changing this.

**Comment by lrhn** [comment]
The current language specification says that the return type of setters *is* `void`. It's the only type you can write, and if you don't write any type, it's still `void`.
Any attempt to make such a `void` value useful is going to be resisted (so no, tearing off a setter will not give you access to the return value, that would invalidate a lot of existing code that uses `=>` *knowing* that the result is unreachable. 
The returned value is not accessible now, and I'd vote a definite no to getting access to the returned value in the future too.
Even `dart:mirrors` seem to hide the return value of setters, you can only invoke them through `setField`.

With that in mind, it seems like a safe optimization to ignore the return. I'd be perfectly happy if the CFE does that for you.

(I still hope to eventually make all `void foo() => e;` functions return `null` instead of the value of `e`, since I didn't succeed in making the syntax invalid).

**Comment by rakudrama** [comment]
@lrhn Thanks for the clarifications.

**Comment by rakudrama** [comment]
Fixed in dart2js via 711dd8e82e022bb57a40d6473444b1696eb0e991
